_episodes/08-exercise.md

```diff
@@ -15,6 +17,87 @@ Now we have developed a range of skills relating to testing and continuous
+  features][pytest_features] large numbers aren't really that different from
+  other numbers unless you run into overflow errors (one could even argue that
+  testing the negative numbers in [pytest features][pytest_features] is not
+  qualitatively different and rather  to double check). One should be aware of
+  the difference between testing for confidence and necessary feature tests. The
+  former is a convenience that comes at the cost of longer test runs and so it
+  is not always desirable to test redundantly (although certainly better than
+  missing an aspect).
+* **test for edge-/corner-cases** - Test special input or conditions for which a
+  general algorithm needs to be specialised (e.g., NaNs, infinities, overflows,
+  empty input, etc.). We did a [whole episode][edge_cases] on this.
+* **test for failures** - This is part of feature testing but important enough to
+  mention explicitly: The conditions under which your code fails are part of
+  your interface and need to be tested. The user (that probably includes
+  yourself) might rely on a raised exception or returned default value in the
+  case of failure. Make sure they can and think of all the cases that your
+  current approach cannot handle. Any changes in these (even those for the
+  better) are changes of the interface and should appear intentionally. This was
+  discussed in [pytest features][pytest_features].
+* **fuzzy testing** - This is broader than testing for failures; if you have
+  unexperienced or even malicious users of your project, they might run your
+  code with inputs or under conditions that do not make any sense at all and are
+  almost impossible to predict. Fuzzy testing is a strategy where you let the
+  computer run your code with random input (sometimes down to the bit level) and
+  make sure that not even the most far-fetched input can break your code. There
+  are libraries for that, so you don't have to set up all the boilerplate
+  yourself.
+* **regression test** - After you have found a bug, you can write a test reproducing
+  the precise conditions under which the bug appeared. Once you fixed it, your
+  test will work fine and if a later change risks introducing this bug again, you
+  can rest assured that it will be immediately signalled by a failing test.
+* **test as a reminder** - In most contemporary test frameworks, you can mark a test
+  as an "expected failure". Such tests are run during your standard test runs
+  but the test framework will complain if they don't fail. This can be a
+  convenient way of marking a to-do or a known bug that you don't have time to
+  fix at the moment. It will preserve your precise intention, e.g., the precise
+  conditions of the bug in code form and it might be an important information if
+  a bug disappeared unexpectedly. Maybe another code change had an effect you
+  did not intend?
+* **test for fixing an external interface** - You can even test code did not write
```

Comment:
Worth mentioning that this is relatively controversial and that having a large battery of tests of other people's code is a code smell (in nicer terms)?

Reply:
Yes, I agree. In a flat list like this, it looks like having the same status as other forms of testing which is definitely not true. Should be sparse and only where really needed. I would however say that you are not at all testing other people's code (in terms of functionality) in that you definitely shouldn't have 100 tests that numpy can correctly add two arrays but you automate the detection of interface changes. Your tests should be designed for that purpose and that purpose alone and they will automatically be sparse and effective that way.